pkg/codegen/hcl2/binder_schema.go

```diff
@@ -43,6 +47,69 @@ func NewPackageCache() *PackageCache {
 	}
 }
 
+func (c *PackageCache) getPackageSchema(name string) (*packageSchema, bool) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	schema, ok := c.entries[name]
+	return schema, ok
+}
+
+// loadPackageSchema loads the schema for a given package by loading the corresponding provider and calling its
+// GetSchema method.
+//
+// TODO: schema and provider versions
+func (c *PackageCache) loadPackageSchema(host plugin.Host, name string) (*packageSchema, error) {
+	if s, ok := c.getPackageSchema(name); ok {
+		return s, nil
+	}
+
+	provider, err := host.Provider(tokens.Package(name), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	schemaBytes, err := provider.GetSchema(0)
+	if err != nil {
+		return nil, err
+	}
+
+	var spec schema.PackageSpec
+	if err := jsoniter.Unmarshal(schemaBytes, &spec); err != nil {
+		return nil, err
+	}
+
+	pkg, err := schema.ImportSpec(spec, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resources := map[string]*schema.Resource{}
+	for _, r := range pkg.Resources {
+		resources[canonicalizeToken(r.Token, pkg)] = r
+	}
+	functions := map[string]*schema.Function{}
+	for _, f := range pkg.Functions {
+		functions[canonicalizeToken(f.Token, pkg)] = f
+	}
+
+	schema := &packageSchema{
+		schema:    pkg,
+		resources: resources,
+		functions: functions,
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if s, ok := c.entries[name]; ok {
```

Comment:
Would it make sense to do the existence check at the beginning of the method and return early if found?

Reply:
We already do: https://github.com/pulumi/pulumi/pull/4661/files#diff-96a8135e5523ef95913cdabca5214134R63-R65